File: historyscan-back/src/test/java/io/sch/historyscan/domain/contexts/analysis/clocrevisions/filesystem/FileSystemTreeTest.java
package io.sch.historyscan.domain.contexts.analysis.clocrevisions.filesystem;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.springframework.core.io.ClassPathResource;

import java.io.IOException;
import java.util.stream.Stream;

import static io.sch.historyscan.domain.contexts.analysis.clocrevisions.filesystem.FileSystemNodeSerializerUtils.serializeExpectedRoot;
import static org.assertj.core.api.Assertions.assertThat;

class FileSystemTreeTest {
    public static Stream<Arguments> should_load_file_system_tree_from_disk_params() {
        return Stream.of(
                Arguments.of("domain", "domain"),
                Arguments.of("/", "theglobalproject"),
                Arguments.of("", "theglobalproject")
        );
    }

    @ParameterizedTest
    @MethodSource("should_load_file_system_tree_from_disk_params")
    void should_load_file_system_tree_from_disk(String rootFolderName, String expectedRootTestCase) throws IOException {
        final RootFolder rootFolder = RootFolder.of(rootFolderName, "theglobalproject");
        var fsTree = new FileSystemTree(rootFolder);
        var codebasesResource = new ClassPathResource("codebases/theglobalproject");

        fsTree.createFrom(new CodeBaseFile(codebasesResource.getFile(), rootFolder));

        final var expectedRoot = serializeExpectedRoot(expectedRootTestCase);
        assertThat(fsTree)
                .extracting(FileSystemTree::getRoot)
                .isEqualTo(expectedRoot);
    }
}

File: historyscan-back/src/test/kotlin/io/sch/historyscan/domain/contexts/analysis/clocrevisions/filesystem/FileSystemTreeTest.kt
package io.sch.historyscan.domain.contexts.analysis.clocrevisions.filesystem

import io.sch.historyscan.domain.contexts.analysis.clocrevisions.filesystem.RootFolder.Companion.of
import org.assertj.core.api.Assertions
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.springframework.core.io.ClassPathResource
import java.io.IOException
import java.util.stream.Stream

internal class FileSystemTreeTest {
    @ParameterizedTest
    @MethodSource("should_load_file_system_tree_from_disk_params")
    @Throws(IOException::class)
    fun should_load_file_system_tree_from_disk(rootFolderName: String, expectedRootTestCase: String) {
        val rootFolder = of(rootFolderName, "theglobalproject")
        val fsTree = FileSystemTree(
            rootFolder
        )
        val codebaseResource = ClassPathResource("codebases/theglobalproject")
        val codebasesResource = ClassPathResource("codebases")

        fsTree.createFrom(
            CodeBaseFile(
                codebaseResource.file, rootFolder,
                codebasesResource.file.path
            )
        )

        val expectedRoot = FileSystemNodeSerializerUtils.serializeExpectedRoot(expectedRootTestCase)
        Assertions.assertThat(fsTree)
            .extracting(FileSystemTree::rootFolder)
            .isEqualTo(expectedRoot)
    }

    companion object {
        @JvmStatic
        fun should_load_file_system_tree_from_disk_params(): Stream<Arguments> {
            return Stream.of(
                Arguments.of("domain", "domain"),
                Arguments.of("mycompany/theproject/domain", "domain-with-path"),
                Arguments.of("/", "theglobalproject"),
                Arguments.of("", "theglobalproject")
            )
        }
    }
}